### DISCUSSION TITLE
能否改进接口的异步调用和限速调用

### DISCUSSION THREAD

**Discussion by acaly** [howmanybots/onebot#94] [想法 / 建议]
按目前的标准，从客户端的角度来看，异步调用有一些不方便使用的地方：

1. 无法得知结果。
2. “另外，有一些接口本身就是异步执行的”。服务器端对非`_async`调用返回异步结果可能会导致在客户端产生预料之外的错误。
3. 每个接口都存在同步和异步两个版本，在处理路由的时候会比较复杂。

我个人的一些想法：
1. 增加一个查询异步调用结果的API，在异步调用返回时同时给出一个token，使用这个token向查询异步API来查之前进行的异步调用的结果。在不需要得知结果的情况下也可以通过参数来告知服务器，这样服务器就不需要返回token并记录异步执行结果了。（也可以默认不给出token，只在被参数明确指定需要查询异步结果的情况下返回token。）
2. 除非被明确指定，禁止API产生async结果。如果不能同步完成则返回错误。另外应该禁止服务器在WebSocket环境下产生async结果（因为连接无论如何都在保持状态）。
3. 异步和限速可以使用参数来指定，而不是在API名后面加后缀。这个参数可以考虑和上面1中的参数合并。

另外，如果我没有理解错的话，异步调用主要是为了限速调用服务的，非限速的异步调用似乎没有实际意义。

**Comment by stdrc** [comment]
其实异步接口这玩意本身就是个特定历史条件的产物，当时的 HTTP server 用的是单线程模型，但调用 CQ API 又是同步阻塞的，于是搞了个线程池来执行异步任务，现在的 OneBot 实现应该不需要受限于这个问题，于是根本不需要异步 API，异步这件事应该在客户端做
